app/src/main/java/de/westnordost/streetcomplete/data/osm/upload/SplitSingleWayUpload.kt

```diff
@@ -0,0 +1,301 @@
+        val wayIdBefore = relation.members.findPrevious(indexInRelation) { it.type == WAY }?.ref
+        val wayBefore = wayIdBefore?.let { osmDao.getWay(it) }
+        if (wayBefore != null) {
+            if (isAfterWayInChain(wayBefore)) return true
+            if (isBeforeWayInChain(wayBefore)) return false
+        }
+
+        val wayIdAfter = relation.members.findNext(indexInRelation+1) { it.type == WAY }?.ref
+        val wayAfter = wayIdAfter?.let { osmDao.getWay(it) }
+        if (wayAfter != null) {
+            if (isBeforeWayInChain(wayAfter)) return true
+            if (isAfterWayInChain(wayAfter)) return false
+        }
+
+        return null
+    }
+
+    private fun Relation.fetchViaNodeIds(relationType: String): List<Long>? {
+        val via = findVia(relationType) ?: return null
+        return when (via.type) {
+            WAY -> osmDao.getWay(via.ref)?.nodeIds?.firstAndLast()
+            NODE -> listOf(via.ref)
+            else -> null
+        }
+    }
+}
+
+/** data class that carries the information for one split to perform on a random position on a way.
+ *  So, same as SplitPolylineAtPosition, but additionally with the index of the split in the way. */
+private sealed class SplitWay : Comparable<SplitWay> {
+    abstract val pos: LatLon
+    protected abstract val index: Int
+    protected abstract val delta: Double
+
+    /** sort by index, then delta, ascending. The algorithm relies on this order! */
+    override fun compareTo(other: SplitWay): Int {
+        val diffIndex = index - other.index
+        if (diffIndex != 0) return diffIndex
+
+        val diffDelta = delta - other.delta
+        return diffDelta.sign.toInt()
+    }
+}
+
+private data class SplitWayAtPoint(override val pos: LatLon, public override val index: Int) : SplitWay() {
+    override val delta get() = 0.0
+}
+
+private data class SplitWayAtLinePosition(
+    val pos1: LatLon, val index1: Int,
+    val pos2: LatLon, val index2: Int,
+    public override val delta: Double) : SplitWay() {
+    override val index get() = index1
+    override val pos: LatLon get() {
+        val line = listOf(pos1, pos2)
+        return SphericalEarthMath.pointOnPolylineFromStart(
+            line,
+            SphericalEarthMath.distance(line) * delta
+        )!!
+    }
+}
+
+/** creates a SplitWay from a SplitLineAtPosition, given the nodes of the way. So, basically it
+ *  simply finds the node index/indices at which the split should be made.
+ *  If the way changed significantly in the meantime, it will throw an ElementConflictException */
+private fun SplitPolylineAtPosition.toSplitWay(positions: List<LatLon>): SplitWay {
+    return when(this) {
+        is SplitAtPoint -> toSplitWay(positions)
+        is SplitAtLinePosition -> toSplitWay(positions)
+    }
+}
+
+private fun SplitAtPoint.toSplitWay(positions: List<LatLon>): SplitWayAtPoint {
+    // could be several indices, for example if the way has the shape of an 8.
```

Comment:
Maybe then the split should be done at *every* node at that position. Same for lines.

Reply:
So, the question is how this should behave for ways that look for example like this:
![Untitled](https://user-images.githubusercontent.com/4661658/63642545-0a947500-c6c1-11e9-9894-127963385e98.png)

If the user clicks on the point where the way intersects itself to split it, should the app split the way at both nodes at that position or just the first?

Reply:
I decided that it should split the way in this example into three pieces. The implementation was rather simple, and that it what a user would expect.